src/binary-encoding.adoc

```diff
@@ -2298,16 +2298,302 @@ considered "shapes" rather than types because while their encoding is always sta
 
 === Encoding E-expressions With Multiple Arguments
 
-TODO
+If a macro has more than one parameter, the E-expression arguments corresponding to each parameter are encoded one
+after the other moving from left to right.
 
-=== Grouped Parameter Encodings
+.Figure {counter:figure}: Definition of macro `_foo_` at address `_0_`
+[%unbreakable,source]
+----
+(macro foo             // Macro name
+  [                    // Parameters
+    (a string!),
+    (b compact_symbol!),
+    (c uint16!)
+  ]
+  /* ... */            // Body (elided)
+)
+----
+
+.Figure {counter:figure}: Encoding of E-expression for macro with multiple parameters: `_(:0 "hello" baz 512)_`
+[%unbreakable,source]
+----
+┌──── The opcode is less than 0x40, so it is an E-expression invoking the macro at
+│     address 0, `foo`. `foo`'s first parameter is a string, so an opcode follows.
+│
+│  ┌──── Opcode 0x85 indicates a 5-byte string. 5 UTF-8 bytes follow.
+│  │
+│  │                 ┌──── `foo`'s second parameter is a compact_symbol, so a `FlexSym` follows.
+│  │                 │     FlexSym -3: 3 bytes of UTF-8 text follow.
+│  │                 │
+│  │                 │           ┌──── `foo`'s third parameter is a uint16, so a 2-byte
+│  │                 │           │     2-byte `FixedUInt` follows.
+│  │                 │           │     FixedUInt: 512
+│  │  h  e  l  l  o  │   b  a  z │
+00 85 68 65 6C 6C 6F FD 62 61 7A 00 20
+      └──────┬─────┘    └───┬──┘
+        UTF-8 bytes    UTF-8 bytes
+----
+
+[[argument_encoding_bitmap]]
+=== Argument Encoding Bitmap (AEB)
+
+The examples in previous sections have only shown how to encode invocations of macros which have either no parameters
+at all (aka _constants_) or whose parameters all have
+link:macros-by-example.adoc#exactly_one[a cardinality of exactly-one].
+
+If a macro has any parameters with a cardinality of zero-or-one (`?`), zero-or-more (`*`), or one-or-more (`+`),
+then E-expressions invoking that macro will begin with an _argument encoding bitmap_ (AEB).
+An AEB is a series of bits that correspond to a macro parameter and communicate additional
+information about how the arguments corresponding to that parameter have been encoded in the current E-expression.
+In particular, the AEB indicates whether a parameter that accepts `(:void)` has any arguments at all, and how a
+grouped parameter's arguments have been delimited.
+
+The number of bits allotted to each parameter is determined by its cardinality, as shown in the table below; each
+parameter can have 0, 1, or 2 bits.
+
+[cols="^.^2a,^.^2a,^.^2a,^.^1a,^.^1a,.^6a"]
+|===
+|Grouping Mode|Cardinality|Example parameter signature|Number of bits|Bit(s) value|Encoding
+
+.6+|Ungrouped
+|Exactly-one
+|`(x int!)`
+|0
+|_n/a_
+|One expression
+
+.2+|Zero-or-one
+.2+|`(x int?)`
+.4+|1
+|`0`
+<|No expression; equivalent to `(:void)`
+
+|`1`
+<|One expression
+
+.2+|Zero-or-more
+.2+|`(x int*)`
+|`0`
+<|No expression; equivalent to `(:void)`
+
+|`1`
+<|One expression
```

Comment:
I think we've (or I've) forgotten either a grammar constraint or an encoding solution, b/c I don't know how to encode a non-unary tagless parameter like `(x uint8+)`, since their "one expression" can't be anything but one value.

Reply:
Yeah, at the moment `uint8+` is not forbidden, but it's identical to `uint8` in both encoding and "expansion." We should probably forbid it to avoid confusion. Assuming there's consensus on that, I'll make the change in a future revision.